src/python/pants/backend/python/subsystems/lambdex.py

```diff
@@ -20,6 +29,34 @@ class Lambdex(PythonToolBase):
     default_lockfile_resource = ("pants.backend.python.subsystems", "lambdex.lock")
     lockfile_rules_type = LockfileRules.SIMPLE
 
+    layout = EnumOption(
+        default=LambdexLayout.LAMBDEX,
+        help=softwrap(
+            """
+            Explicitly control the layout used for `python_awslambda` and
+            `python_google_cloud_function` targets. This option exists for the transition from
+            Lambdex-based layout to the plain zip layout, as recommended by cloud vendors.
+            """
+        ),
+    )
+
+    def warn_for_layout(self, target_alias: str) -> None:
+        if self.options.is_default("layout"):
+            warn_or_error(
+                "2.18.0.dev0",
+                f"using the Lambdex layout by default for `{target_alias}` targets",
+                softwrap(
+                    f"""
+                    To prepare for the transition to the new layout for `{target_alias}` targets in
```

Comment:
This is not quite right - we want people to set it to "zip" (so they can be sure that is going to work for them). "lambdex" is the escape hatch if "zip" doesn't work.   So the message should express that.

Reply:
Ah, I think I might've misinterpreted the suggestion on #19032 (not a problem from my side, happy to keep iterating).

**Option A**: I was thinking we'd be doing something like this:

| release | supports lambdex? | supports zip? | default layout | deprecation warnings |
|---|---|---|---|---|
| 2.17 | ✅ | ❌ | lambdex | if `layout = "lambdex"` is implicit, tell people to set it to prep for 2.18 |
| 2.18 | ✅ | ✅ | zip | if `layout = "lambdex"` is set at all, tell people to remove it and switch to `zip` |
| 2.19 | ❌ | ✅ | zip | none, migration over (or maybe just about removing the `[lambdex]` section entirely) |

That is, this patch focused on just what we need to cherry-pick back to 2.17, but doesn't actually change the FaaS-packaging functionality itself, so having a `zip` option and telling people to use it will be misleading. 

The next step for #18879 is #19076, which implements the 2.18 behaviour (including adding a `zip` option and flipping the defaults over), but that's separate because it shouldn't be cherry picked.

(NB. #19022 / 2bc078b0252931ccd5140432a04f3328a0d32e3f missed the 2.17 branch, even without the revert #19071.)

**Option B**: I could also imagine something like:

| release | supports lambdex? | supports zip? | default layout | deprecation warnings |
|---|---|---|---|---|
| 2.17 | ✅ | ✅  | lambdex | if `layout = "lambdex"` is implicit, tell people to set it: recommend `zip`, but allow `lambdex` if they have to |
| 2.18 | ✅ | ✅ | zip | if `layout = "lambdex"` is set at all, tell people to remove it and switch to `zip` |
| 2.19 | ❌ | ✅ | zip | none, migration over (or maybe just about removing the `[lambdex]` section entirely) |

But, that would require cherry-picking _all_ the new FaaS-packaging functionality to the 2.17 branch, and that feels a bit risky, to me.

**Option C**: option B but without any cherry-picks: increment each version by one (i.e. start the deprecations in 2.18 and finish in 2.20).

I was thinking and implementing towards option A. Were you thinking of one of these, or something else?

Reply:
Ah yes, I was imagining Option B. I think it's still OK to cherry-pick to 2.17, since we haven't done an RC from that branch yet. But I can also go with Option C just as well if you think that's better. And yes, in Option B/2.19 (Option C/2.20) we remove the lambdex subsystem entirely.

Option A doesn't really add any value - we only need to call attention to any of this to start the transition to `zip`, and users can't do anything towards that in 2.17 in option A.

Reply:
Okay, I'll go with option B. I was thinking option A because that let's us finish the deprecation cycle quicker than option C by forcing 2.17 users to prepare for the new default in 2.18, with less cherry-picking than option B, but if you're happy with cherry-picking something the size of #19076, then I'll do that. Time for me to rearrange the PRs again!

Thanks for walking me through this.

Reply:
Option A doesn't really give users any extra preparation though, because they can't try `zip`, so all it does is warn them that a change is gonna come, but not in a way that let's them do anything about it.

Reply:
So yeah, I think cherry-picking into 2.17.x is fine (not 2.16.x though!)

Reply:
I've incorporated the relevant parts of this into https://github.com/pantsbuild/pants/pull/19076, so I'll close.